File: MainApp/src/main/java/com/sleticalboy/learning/components/provider/StoreProvider.java
package com.sleticalboy.learning.components.provider;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.net.Uri;
import android.util.Log;

public class StoreProvider extends ContentProvider {

    private static final String TAG = "StoreProvider";

    private static final String AUTHORITY = "com.sleticalboy.dailywork.store";
    private static final UriMatcher URI_MATCHER = new UriMatcher(UriMatcher.NO_MATCH);

    static {
        URI_MATCHER.addURI(AUTHORITY, "/cache", 0);
        URI_MATCHER.addURI(AUTHORITY, "/pair", 1);
    }

    public StoreProvider() {
    }

    @Override
    public int delete(Uri uri, String selection, String[] selectionArgs) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public String getType(Uri uri) {
        return null;
    }

    @Override
    public Uri insert(Uri uri, ContentValues values) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public boolean onCreate() {
        // 此方法会比 Application 的 onCreate() 方法先执行
        Log.d(TAG, "onCreate() start --->");
        StoreManager.init(getContext());
        return true;
    }

    @Override
    public Cursor query(Uri uri, String[] projection, String selection,
                        String[] selectionArgs, String sortOrder) {
        final MatrixCursor cursor = new MatrixCursor(projection);
        int value = 0;
        final String table = uri.getLastPathSegment();
        String sql = "select mac from device_list where mac=? and status";
        if ("cache".equals(table)) {
            sql += "=0";
        } else if ("pair".equals(table)) {
            sql += ">=3";
        } else {
            cursor.addRow(new Object[]{value});
            return cursor;
        }
        final Cursor c = StoreManager.get().readableDb().rawQuery(sql, projection);
        if (c != null) {
            value = c.getCount() != 0 ? 1 : 0;
            c.close();
        }
        cursor.addRow(new Object[]{value});
        return cursor;
    }

    @Override
    public int update(Uri uri, ContentValues values, String selection,
                      String[] selectionArgs) {
        throw new UnsupportedOperationException("Not yet implemented");
    }
}

File: MainApp/src/main/java/com/sleticalboy/learning/components/provider/StoreProvider.kt
package com.sleticalboy.learning.components.provider

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.MatrixCursor
import android.net.Uri
import android.util.Log

class StoreProvider : ContentProvider() {

    companion object {
        private const val TAG = "StoreProvider"
        private const val AUTHORITY = "com.sleticalboy.dailywork.store"
        private val URI_MATCHER = UriMatcher(UriMatcher.NO_MATCH)

        init {
            URI_MATCHER.addURI(AUTHORITY, "/cache", 0)
            URI_MATCHER.addURI(AUTHORITY, "/pair", 1)
        }
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        throw UnsupportedOperationException("Not yet implemented")
    }

    override fun getType(uri: Uri): String? {
        return null
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        throw UnsupportedOperationException("Not yet implemented")
    }

    override fun onCreate(): Boolean {
        // 此方法会比 Application 的 onCreate() 方法先执行
        Log.d(TAG, "onCreate() start --->")
        StoreManager.Companion.init(context)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?,
                       selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        val cursor = MatrixCursor(projection)
        var value = 0
        val table = uri.lastPathSegment
        var sql = "select mac from device_list where mac=? and status"
        sql += when (table) {
            "cache" -> {
                "=0"
            }
            "pair" -> {
                ">=3"
            }
            else -> {
                cursor.addRow(arrayOf<Any>(value))
                return cursor
            }
        }
        val c: Cursor = StoreManager.get().readableDb().rawQuery(sql, projection)
        value = if (c.count != 0) 1 else 0
        c.close()
        cursor.addRow(arrayOf<Any>(value))
        return cursor
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?,
                        selectionArgs: Array<String>?): Int {
        throw UnsupportedOperationException("Not yet implemented")
    }
}